### PR TITLE
Terminate dataloader workers properly when parent process is SIGKILL'ed

### PR DESCRIPTION
**Pull request by yf225**
Reopening https://github.com/pytorch/pytorch/pull/6606 with fix for `TEST_CUDA` import issue on Windows and improvement to how we wait for manager exit in `test_manager_unclean_exit`. Need to loop test the Windows CI multiple times to make sure this actually fixes the CUDA OOM issue.

### CODE REVIEW

File: test/test_dataloader.py [hunk 1]
```diff
@@ -8,15 +8,26 @@
...
+import subprocess
 from torch import multiprocessing
 from torch.utils.data import Dataset, TensorDataset, DataLoader, ConcatDataset
 from torch.utils.data.dataset import random_split
-from torch.utils.data.dataloader import default_collate, ExceptionWrapper
+from torch.utils.data.dataloader import default_collate, ExceptionWrapper, MANAGER_STATUS_CHECK_INTERVAL
 from common import TestCase, run_tests, TEST_NUMPY, IS_WINDOWS
-from common_nn import TEST_CUDA
 
+TEST_CUDA = torch.cuda.is_available()
```
Comment (yf225) [inline test/test_dataloader.py]: If we import `TEST_CUDA` from `common_nn` multiple times during the execution of the test, `TEST_CUDNN = TEST_CUDA and torch.backends.cudnn.is_acceptable(torch.cuda.FloatTensor(1))` in `common_nn` will also be executed multiple times and CUDA on Windows doesn't seem to like it. This is an attempt to find a workaround for the issue.

Comment (ezyang) [reply test/test_dataloader.py]: OK, can this live as a code comment, so someone doesn't "helpfully" refactor `TEST_CUDA` back into `common_nn`?

File: test/test_dataloader.py [hunk 2]
```diff
@@ -8,15 +8,26 @@
...
 
-JOIN_TIMEOUT = 17.0 if IS_WINDOWS else 4.5
+# We need spawn start method for test_manager_unclean_exit
+if sys.version_info[0] == 3:
```
Comment (ezyang) [inline test/test_dataloader.py]: Why is the version test necessary here?

File: torch/utils/data/dataloader.py [hunk 3]
```diff
@@ -30,6 +37,37 @@ def __init__(self, exc_info):
 _use_shared_memory = False
 r"""Whether to use shared memory in default_collate"""
 
+MANAGER_STATUS_CHECK_INTERVAL = 5.0
+
+if IS_WINDOWS:
+    class ManagerWatchdog(object):
```
Comment (ezyang) [inline torch/utils/data/dataloader.py]: Here would be a good place for a comment discussing why the alternate ManagerWatchdog implementation is needed on Windows.

**Review by ezyang — APPROVED** [review]

File: test/test_dataloader.py (hunk 2 above)
Comment (yf225) [reply test/test_dataloader.py]: `multiprocessing` doesn't have `set_start_method()` on Python 2.7 and can only use the `fork` start method on Linux. I can add a comment to explain it.

File: test/test_dataloader.py [hunk 4]
```diff
@@ -8,15 +8,31 @@
...
-from torch.utils.data.dataloader import default_collate, ExceptionWrapper
+from torch.utils.data.dataloader import default_collate, ExceptionWrapper, MANAGER_STATUS_CHECK_INTERVAL
 from common import TestCase, run_tests, TEST_NUMPY, IS_WINDOWS
-from common_nn import TEST_CUDA
 
+# We cannot import TEST_CUDA from common_nn here, because if we do that,
+# the TEST_CUDNN line from common_nn will be executed multiple times
+# as well during the execution of this test suite, and it will cause
+# CUDA OOM error on Windows.
+TEST_CUDA = torch.cuda.is_available()
```
Comment (apaszke) [inline test/test_dataloader.py]: Could we perhaps made those properties lazy? You could make them objects with a `__bool__` method that does the check only when called for the first time.

**Pull request merged**